### DISCUSSION TITLE
Unit testing not working as expected

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#10275] [Using MicroPython] [answered]
Hi,
I'm trying to do unittests but I cannot get it to work properly... I tried to do a simple test run...
On unix I have a working micropython instance. I installed `unittest` and `unittest-discover` successfully using `mip`:
```
micropython -m mip install unittest
micropython -m mip install unittest-discover
```
My simple unittest testing directory looks like:
```
├── app
│   ├── __init__.py
│   └── calculations.py
└── tests
    ├── __init__.py
    ├── test_calculations.py
```

Running an individual test (without a **tests** dir) works but when running `micropython -m unittest` I get:
```
Traceback (most recent call last):
  File "unittest/__main__.py", line 142, in <module>
  File "unittest/__main__.py", line 123, in discover_main
  File "unittest/__main__.py", line 89, in _discover
  File "unittest/__main__.py", line 42, in _run_all_in_dir
  File "unittest/__main__.py", line 46, in _run_all_in_dir
  File "fnmatch.py", line 50, in fnmatch
  File "fnmatch.py", line 86, in fnmatchcase
  File "fnmatch.py", line 60, in _compile_pattern
  File "fnmatch.py", line 128, in translate
  File "fnmatch.py", line 135, in re_escape
AttributeError: 'module' object has no attribute 'sub'
```

Is there any example how to get unittest to work when bundling all tests inside another tests directory?
Thanks!

FYI:
**app/calculations.py** looks like:
```
class Calculations:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def get_sum(self):
        return self.a + self.b

    def get_difference(self):
        return self.a - self.b

    def get_product(self):
        return self.a * self.b

    def get_quotient(self):
        return self.a / self.b
```

and **tests/test_calculations.py** looks like:
```
import unittest
from app.calculations import Calculations

class TestCalculations(unittest.TestCase):

    def test_sum(self):
        calculation = Calculations(8, 2)
        self.assertEqual(calculation.get_sum(), 10, 'The sum is wrong.')

    def test_diff(self):
        calculation = Calculations(8, 2)
        self.assertEqual(calculation.get_difference(), 6, 'The difference is wrong.')

    def test_product(self):
        calculation = Calculations(8, 2)
        self.assertEqual(calculation.get_product(), 16, 'The product is wrong.')

    def test_quotient(self):
        calculation = Calculations(8, 2)
        self.assertEqual(calculation.get_quotient(), 4, 'The quotient is wrong.')

if __name__ == '__main__':
    unittest.main()
```

Both **__init__.py** files are empty

**Comment by jimmo** [comment, marked as answer]
> AttributeError: 'module' object has no attribute 'sub'

@saraverbeecke The problem is that fnmatch is using re.sub (see https://github.com/micropython/micropython-lib/blob/master/python-stdlib/fnmatch/fnmatch.py#L135) but this is not available ~by default on the unix build~ on whichever build you're using.

But in the meantime, you can enable it in your build by adding

```c
#define MICROPY_PY_URE_SUB (1)
```

in either unix/mpconfigport.h (or your mpconfigvariant.h if you have one).

**Reply by jimmo** [reply]
Wait... Unix should already have that enabled (as it enables "extra features" by default). Where are you getting the `micropython` binary from?

Perhaps you have an older version. If you build from the latest source it should work. (This is what I was using when I wrote unittest-discover).

**Reply by saraverbeecke** [reply]
@jimmo thanks, indeed it works with the latest unix version. My bad! For my quick test I used the unix micropython version installed via `brew install micropython` and that did not include the latest release apparently. 
Is there any way to get the test coverage as well using MicroPython's version of unittest + unitt-discover. I did not find a "coverage" module in the micropython-lib repo...

**Reply by jimmo** [reply]
Unfortunately coverage is not possible with MicroPython at the moment.

It is technically feasible to get coverage working using sys.settrace, I was actually looking at this recently for fun but not likely something I'll be able to put any time into anytime soon.

But in the meantime the best thing to do is to _also_ run your unit tests in CPython (with stubs as necessary for MicroPython APIs) and do the coverage there.
